### PR TITLE
Extend naming scheme to mac address policy and introduce NAMING_STABLE_VIRTUAL_MACS

### PR DESCRIPTION
**Pull request by keszybz**
We should have done this when originally introducing the change. Doing this now will not magically
fix things, because the change is already out in the wild, but it is still useful for distributions which are
only upgrading from v240 or earlier and the users are not depending on the changed behaviour.
Doing this now also establishes a pattern we can follow the next time a naming/mac/whatever change
comes up.

### CODE REVIEW

File: src/libsystemd-network/dhcp-identifier.c [hunk 1]
```diff
@@ -190,7 +190,7 @@ int dhcp_identifier_set_iaid(
                                 /* device is under renaming */
                                 return -EBUSY;
 
-                        name = net_get_name(device);
+                        name = net_get_name_persistent(device);
```
Comment (yuwata) [inline src/libsystemd-network/dhcp-identifier.c:193]: This should be included in the first commit.

File: src/udev/net/link-config.c [hunk 2]
```diff
@@ -322,6 +322,7 @@ static int get_mac(sd_device *device, MACAddressPolicy policy, struct ether_addr
                 if (r < 0)
                         return log_device_warning_errno(device, r, "Could not generate persistent MAC: %m");
 
+                log_device_debug(device, "Using generated persistent MAC address");
```
Comment (yuwata) [inline src/udev/net/link-config.c:327]: Not a matter, but I'd like this to be included in the second commit.

**Comment by yuwata** [comment]
So, are you planning to backport this change into v241 and v242?

**Comment by keszybz** [comment]
Updated with the requested changes. Now there are three commits.

> So, are you planning to backport this change into v241 and v242?

I was wondering about this. I think this is not suitable for the stable branches, because too much time has passes the release. But I definitely think that it makes sense for distributions which are now updating to v241+ to backport those patches.

**Comment by yuwata** [comment]
LGTM.

**Comment by fbuihuu** [comment]
First thanks @keszybz @yuwata for looking into this.

But I'm not sure how distros are supposed to make use of the naming scheme in general. 

In my understanding the naming scheme version should be fixed at package installation time  and should remain constant across package upgrades hence the device name/mac are also preserved. But also new installations should use the latest version.

Is that correct ?

If so fixing the default naming scheme through a build option doesn't sound appropriate. Or maybe we're supposed to ship a drop-in for udev which would fix the naming scheme version via `NET_NAMING_SCHEME`  environment variable ?

**Comment by keszybz** [comment]
There are two ways:
1. Set the naming scheme during package build. This is useful for distros which do release, like Fedora, RHEL, Debian, etc. The naming scheme would be only bumped when doing a new major release of the distro. When this is done, the naming scheme only changes during distro version upgrades, which often bring significant other changes anyway, and in particular kernel changes.
2. Save the naming scheme during installation to the boot loader configuration, so that it gets set on the kernel command line. Then the naming scheme would be fixed for the lifetime of the installation.

>  Or maybe we're supposed to ship a drop-in for udev which would fix the naming scheme version via NET_NAMING_SCHEME environment variable

That would be an option too. I didn't think about this approach, but it should work.

As far as I know, no distribution takes any of those approaches so far.

**Comment by fbuihuu** [comment]
> There are two ways:
> 
>     1. Set the naming scheme during package build. This is useful for distros which do release, like Fedora, RHEL, Debian, etc. The naming scheme would be only bumped when doing a new major release of the distro. When this is done, the naming scheme only changes during distro version upgrades, which often bring significant other changes anyway, and in particular kernel changes.

Even in this case, I doubt that users (of desktop systems) read (and understand) all the changes that are brought by upgrades. But they will be most likely unhappy when their network settings will break IMHO. This is especially true for people who are running a rolling distro, as you could see.

Backward compatibility breakages should be avoided especially since it seems it can easily be avoided now by fixing the naming scheme at package installation.

> 
>     2. Save the naming scheme during installation to the boot loader configuration, so that it gets set on the kernel command line. Then the naming scheme would be fixed for the lifetime of the installation.
> 
> 
> > Or maybe we're supposed to ship a drop-in for udev which would fix the naming scheme version via NET_NAMING_SCHEME environment variable
> 
> That would be an option too. I didn't think about this approach, but it should work.

I dunno, both methods sound a bit awkward: passing the naming scheme (or any option actually) through an environment variable is (usually) pretty ugly and tweaking the kernel command line option via the bootloader sound like a pain if not done by the installer.

Maybe another possibility would be to use a private file (in /usr/lib/udev) read by udev that  would contain the naming scheme. It could be easily installed at package installation (or even by meson) and could be overwritten by the user via the kernel command line later.

**Comment by fbuihuu** [comment]
Also I think it can be useful to allow passing an "approximate" naming scheme version. Therefore if  for example systemd version "v242" is installed then the actual naming scheme used by udev would be `NAMING_V241` since `NAMING_V242` doesn't exist currently.

But if for some reasons `NAMING_V242` needs to be created in the future then this scheme would be used instead.

**Comment by keszybz** [comment]
There is no magic bullet. Each choice has its pluses and minuses.

> I dunno, both methods sound a bit awkward: passing the naming scheme (or any option actually) through an environment variable is (usually) pretty ugly and tweaking the kernel command line option via the bootloader sound like a pain if not done by the installer.

The reason to do it this way is to allow renaming to happen very early. Doing it as a boot loader option (or compilation time switch) has the advantage that it takes effect already in the initramfs. But allowing interface renames (even if the interface has a user-specified name), and reading a normal config file in /etc would be an option. I think we should also support this.

> pain if not done by the installer.

Yes, I think this should be done by the installer. It's an installation-time distro-level decision how to handle this.

**Pull request merged**